syntax = "proto3";

package foreverbull.service;

option go_package = "github.com/lhjnilsson/foreverbull/pkg/service/pb";

import "google/protobuf/struct.proto";
import "foreverbull/finance/finance.proto";
import "foreverbull/service/worker.proto";

message GetServiceInfoRequest {
}

message GetServiceInfoResponse {
    Algorithm algorithm = 1;
}

message ConfigureExecutionRequest {
    ExecutionConfiguration configuration = 1;
}

message ConfigureExecutionResponse {
}

message RunExecutionRequest {
}

message RunExecutionResponse {
}

message WorkerRequest {
    string task = 1;
    repeated string symbols = 2;
    foreverbull.finance.Portfolio portfolio = 3;
}

message WorkerResponse {
    string task = 1;
    repeated foreverbull.finance.Order orders = 2;
    optional string error = 3;
}

enum NamespaceRequestType {
    GET = 0;
    SET = 1;
}

message NamespaceRequest {
    string key = 1;
    NamespaceRequestType type = 2;
    optional google.protobuf.Struct value = 3;
}

message NamespaceResponse {
    optional google.protobuf.Struct value = 1;
    optional string error = 2;
}


service Worker {
    rpc GetServiceInfo (GetServiceInfoRequest) returns (GetServiceInfoResponse) {}
    rpc ConfigureExecution (ConfigureExecutionRequest) returns (ConfigureExecutionResponse) {}
    rpc RunExecution (RunExecutionRequest) returns (RunExecutionResponse) {}
}
